File: mysql/proxysql/python/proxysql_monitor/formatter.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Iterable, List, Optional, Pattern, Sequence, Tuple

from .models import BackendRows, ConnectionRow, DigestRow, PingRow, PoolRow, QueryRow
from .queries import BACKEND_PING_MARKER, BACKEND_POOL_MARKER

BOLD = "\x1b[1m"
RED = "\x1b[1;31m"
GREEN = "\x1b[1;32m"
YELLOW = "\x1b[1;33m"
ORANGE = "\x1b[1;38;5;208m"
CYAN = "\x1b[1;36m"
RESET = "\x1b[0m"


@dataclass(frozen=True)
class RenderedView:
    colored: str
    clean: str
    row_count: int


def _parse_int(value: str, field_name: str) -> int:
    try:
        return int(value)
    except ValueError as exc:
        raise ValueError(f"Invalid {field_name}: {value!r}") from exc


def sanitize_text(value: str) -> str:
    """Flatten escaped and literal controls so every record stays on one line."""
    value = re.sub(r"\\[nrt]", " ", value)
    value = "".join(
        " " if ord(character) < 32 or 127 <= ord(character) <= 159 else character
        for character in value
    )
    return re.sub(r" +", " ", value).strip()


def truncate(value: str, width: int) -> str:
    """Truncate a string to a non-negative display width."""
    return value[: max(0, width)]


def compile_user_filter(value: str) -> Pattern[str]:
    """Compile the monitor's comma-as-regex-alternation user filter."""
    expression = value.replace(",", "|") if value else ".*"
    try:
        return re.compile(expression, re.IGNORECASE)
    except re.error as exc:
        raise ValueError(f"Invalid user filter regex: {exc}") from exc


def parse_connections(lines: Sequence[str]) -> Tuple[ConnectionRow, ...]:
    rows: List[ConnectionRow] = []
    for line in lines:
        if not line:
            continue
        fields = line.split("\t")
        if len(fields) != 5:
            raise ValueError(f"Malformed CONN row: {line!r}")
        rows.append(
            ConnectionRow(
                fields[0], fields[1], fields[2], fields[3],
                _parse_int(fields[4], "connections"),
            )
        )
    return tuple(rows)


def parse_queries(lines: Sequence[str]) -> Tuple[QueryRow, ...]:
    rows: List[QueryRow] = []
    for line in lines:
        if not line:
            continue
        fields = line.split("\t", 6)
        if len(fields) != 7:
            raise ValueError(f"Malformed QUERY row: {line!r}")
        rows.append(
            QueryRow(
                fields[0], fields[1], fields[2], fields[3], fields[4],
                _parse_int(fields[5], "time_ms"), sanitize_text(fields[6]),
            )
        )
    return tuple(rows)


def parse_digests(lines: Sequence[str]) -> Tuple[DigestRow, ...]:
    rows: List[DigestRow] = []
    for line in lines:
        if not line:
            continue
        fields = line.split("\t", 5)
        if len(fields) != 6:
            raise ValueError(f"Malformed DIGEST row: {line!r}")
        rows.append(
            DigestRow(
                fields[0],
                _parse_int(fields[1], "count_star"),
                _parse_int(fields[2], "sum_time"),
                _parse_int(fields[3], "min_time"),
                _parse_int(fields[4], "max_time"),
                sanitize_text(fields[5]),
            )
        )
    return tuple(rows)


def parse_backend(lines: Sequence[str]) -> BackendRows:
    if not lines:
        return BackendRows((), ())
    try:
        pool_index = lines.index(BACKEND_POOL_MARKER)
        ping_index = lines.index(BACKEND_PING_MARKER)
    except ValueError as exc:
        raise ValueError("BACKEND payload is missing section markers.") from exc
    if pool_index >= ping_index:
        raise ValueError("BACKEND section markers are out of order.")

    pool: List[PoolRow] = []
    for line in lines[pool_index + 1 : ping_index]:
        if not line:
            continue
        fields = line.split("\t")
        if len(fields) != 7:
            raise ValueError(f"Malformed BACKEND pool row: {line!r}")
        pool.append(
            PoolRow(
                fields[0], fields[1], fields[2],
                _parse_int(fields[3], "ConnUsed"),
                _parse_int(fields[4], "ConnFree"),
                _parse_int(fields[5], "ConnOK"),
                _parse_int(fields[6], "ConnERR"),
            )
        )

    ping: List[PingRow] = []
    for line in lines[ping_index + 1 :]:
        if not line:
            continue
        fields = line.split("\t", 3)
        if len(fields) != 4:
            raise ValueError(f"Malformed BACKEND ping row: {line!r}")
        success_value = fields[2].strip()
        success: Optional[int] = None
        if success_value and success_value.upper() != "NULL":
            success = _parse_int(success_value, "ping_success_time_us")
        ping.append(PingRow(fields[0], fields[1], success, sanitize_text(fields[3])))
    return BackendRows(tuple(pool), tuple(ping))


def _render(lines: Iterable[Tuple[str, str]], row_count: int) -> RenderedView:
    pairs = list(lines)
    return RenderedView(
        colored="\n".join(colored for colored, _clean in pairs),
        clean="\n".join(clean for _colored, clean in pairs),
        row_count=row_count,
    )


def _backend_status_color(status: str) -> str:
    return {
        "ONLINE": GREEN,
        "SHUNNED": YELLOW,
        "OFFLINE_SOFT": ORANGE,
        "OFFLINE_HARD": RED,
    }.get(status.strip().upper(), RED)


def format_connections(
    rows: Sequence[ConnectionRow],
    previous: Sequence[ConnectionRow],
    *,
    threshold: int,
    user_filter: str,
) -> RenderedView:
    matcher = compile_user_filter(user_filter)
    selected = [row for row in rows if matcher.search(sanitize_text(row.user))]
    old = {
        (row.user, row.client_host, row.server_host, row.schema): row.connections
        for row in previous
    }
    output: List[Tuple[str, str]] = []
    header = (
        f"{'USER':<20} | {'SOURCE (Cli)':<15} | {'BACKEND (Srv)':<28} | "
        f"{'SCHEMA':<20} | {'CONN':<10} | {'DELTA':<10}"
    )
    output.append((BOLD + header + RESET, header))
    total_connections = 0
    total_delta = 0
    for row in selected:
        key = (row.user, row.client_host, row.server_host, row.schema)
        delta = row.connections - old.get(key, row.connections)
        delta_text = f"{delta:+d}" if delta else "0"
        total_connections += row.connections
        total_delta += delta
        clean = (
            f"{truncate(sanitize_text(row.user), 20):<20} | "
            f"{truncate(sanitize_text(row.client_host), 15):<15} | "
            f"{truncate(sanitize_text(row.server_host), 28):<28} | "
            f"{truncate(sanitize_text(row.schema), 20):<20} | "
            f"{row.connections:<10} | {delta_text:<10}"
        )
        color = RED if threshold > 0 and row.connections >= threshold else GREEN
        output.append((color + clean + RESET, clean))
    if total_connections > 0:
        total_delta_text = f"{total_delta:+d}" if total_delta else "0"
        clean = (
            f"{'GLOBAL TOTALS':<20} | {'':<15} | {'':<28} | {'':<20} | "
            f"{total_connections:<10} | {total_delta_text:<10}"
        )
        output.append((CYAN + BOLD + clean + RESET, clean))
    return _render(output, len(selected))


def format_queries(
    rows: Sequence[QueryRow],
    *,
    user_filter: str,
    terminal_width: int = 130,
) -> RenderedView:
    matcher = compile_user_filter(user_filter)
    selected = [row for row in rows if matcher.search(sanitize_text(row.user))]
    query_width = max(20, terminal_width - 97)
    output: List[Tuple[str, str]] = []
    header = (
        f"{'PSID':<8} | {'HG':<4} | {'USER':<15} | {'SOURCE':<15} | "
        f"{'BACKEND':<28} | {'TIME':<9} | {'ACTIVE QUERY':<{query_width}}"
    )
    output.append((BOLD + header + RESET, header))
    for row in selected:
        clean = (
            f"{truncate(sanitize_text(row.session_id), 8):<8} | "
            f"{truncate(sanitize_text(row.hostgroup), 4):<4} | "
            f"{truncate(sanitize_text(row.user), 15):<15} | "
            f"{truncate(sanitize_text(row.client_host), 15):<15} | "
            f"{truncate(sanitize_text(row.server_host), 28):<28} | "
            f"{str(row.time_ms) + 'ms':<9} | "
            f"{truncate(sanitize_text(row.query), query_width):<{query_width}}"
        )
        color = RED if row.time_ms > 1000 else YELLOW if row.time_ms > 500 else CYAN
        output.append((color + clean + RESET, clean))
    return _render(output, len(selected))


def format_digests(
    rows: Sequence[DigestRow],
    *,
    user_filter: str,
    terminal_width: int = 130,
) -> RenderedView:
    del user_filter
    query_width = max(20, terminal_width - 73)
    output: List[Tuple[str, str]] = []
    header = (
        f"{'DIGEST':<18} | {'COUNT':<10} | {'SUM_TIME':<10} | "
        f"{'MIN_TIME':<10} | {'MAX_TIME':<10} | {'QUERY TEXT':<{query_width}}"
    )
    output.append((BOLD + header + RESET, header))
    for row in rows:
        clean = (
            f"{truncate(sanitize_text(row.digest), 18):<18} | {row.count:<10} | "
            f"{row.sum_time:<10} | {row.min_time:<10} | {row.max_time:<10} | "
            f"{truncate(sanitize_text(row.query), query_width):<{query_width}}"
        )
        output.append((CYAN + clean + RESET, clean))
    return _render(output, len(rows))


def format_backend(rows: BackendRows, terminal_width: int = 130) -> RenderedView:
    del terminal_width
    output: List[Tuple[str, str]] = []
    pool_header = (
        f"{'HOSTGROUP':<10} | {'BACKEND HOST':<35} | {'STATUS':<15} | "
        f"{'CONN USED':<11} | {'CONN FREE':<11} | {'CONN OK':<11} | {'CONN ERR':<11}"
    )
    output.append((BOLD + pool_header + RESET, pool_header))
    for row in rows.pool:
        clean = (
            f"{truncate(sanitize_text(row.hostgroup), 10):<10} | "
            f"{truncate(sanitize_text(row.server_host), 35):<35} | "
            f"{truncate(sanitize_text(row.status), 15):<15} | "
            f"{row.conn_used:<11} | {row.conn_free:<11} | "
            f"{row.conn_ok:<11} | {row.conn_err:<11}"
        )
        color = _backend_status_color(row.status)
        output.append((color + clean + RESET, clean))
    ping_header = (
        f"{'HOSTNAME':<35} | {'LAST PING DATETIME':<25} | "
        f"{'SUCCESS (us)':<15} | PING ERROR"
    )
    output.append((BOLD + ping_header + RESET, ping_header))
    for row in rows.ping:
        success = "" if row.success_us is None else str(row.success_us)
        clean = (
            f"{truncate(sanitize_text(row.hostname), 35):<35} | "
            f"{truncate(sanitize_text(row.last_ping), 25):<25} | "
            f"{success:<15} | {sanitize_text(row.error)}"
        )
        normalized_error = row.error.strip().upper()
        color = GREEN if normalized_error in ("", "NULL") else RED
        output.append((color + clean + RESET, clean))
    return _render(output, len(rows.pool) + len(rows.ping))

File: mysql/proxysql/go/internal/formatter/formatter.go
package formatter

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"unicode/utf8"

	"github.com/Fenicce77/fenicceTools/mysql/proxysql/go/internal/model"
	"github.com/Fenicce77/fenicceTools/mysql/proxysql/go/internal/queries"
)

const (
	bold   = "\x1b[1m"
	red    = "\x1b[1;31m"
	green  = "\x1b[1;32m"
	yellow = "\x1b[1;33m"
	orange = "\x1b[1;38;5;208m"
	cyan   = "\x1b[1;36m"
	reset  = "\x1b[0m"
)

func parseInt(value, field string) (int, error) {
	number, err := strconv.Atoi(value)
	if err != nil {
		return 0, fmt.Errorf("invalid %s %q: %w", field, value, err)
	}
	return number, nil
}

func Sanitize(value string) string {
	replacer := strings.NewReplacer(
		`\n`, " ", `\r`, " ", `\t`, " ",
	)
	value = replacer.Replace(value)
	value = strings.Map(func(character rune) rune {
		if character < 32 || (character >= 127 && character <= 159) {
			return ' '
		}
		return character
	}, value)
	return strings.Join(strings.Fields(value), " ")
}

func truncate(value string, width int) string {
	if width <= 0 {
		return ""
	}
	if utf8.RuneCountInString(value) <= width {
		return value
	}
	return string([]rune(value)[:width])
}

func CompileUserFilter(value string) (*regexp.Regexp, error) {
	if value == "" {
		value = ".*"
	} else {
		value = strings.ReplaceAll(value, ",", "|")
	}
	filter, err := regexp.Compile("(?i)" + value)
	if err != nil {
		return nil, fmt.Errorf("invalid user filter regex: %w", err)
	}
	return filter, nil
}

func ParseConnections(lines []string) ([]model.ConnectionRow, error) {
	rows := make([]model.ConnectionRow, 0, len(lines))
	for _, line := range lines {
		if line == "" {
			continue
		}
		fields := strings.Split(line, "\t")
		if len(fields) != 5 {
			return nil, fmt.Errorf("malformed CONN row %q", line)
		}
		connections, err := parseInt(fields[4], "connections")
		if err != nil {
			return nil, err
		}
		rows = append(rows, model.ConnectionRow{
			User: fields[0], ClientHost: fields[1], ServerHost: fields[2],
			Schema: fields[3], Connections: connections,
		})
	}
	return rows, nil
}

func ParseQueries(lines []string) ([]model.QueryRow, error) {
	rows := make([]model.QueryRow, 0, len(lines))
	for _, line := range lines {
		if line == "" {
			continue
		}
		fields := strings.SplitN(line, "\t", 7)
		if len(fields) != 7 {
			return nil, fmt.Errorf("malformed QUERY row %q", line)
		}
		timeMS, err := parseInt(fields[5], "time_ms")
		if err != nil {
			return nil, err
		}
		rows = append(rows, model.QueryRow{
			SessionID: fields[0], Hostgroup: fields[1], User: fields[2],
			ClientHost: fields[3], ServerHost: fields[4], TimeMS: timeMS,
			Query: Sanitize(fields[6]),
		})
	}
	return rows, nil
}

func ParseDigests(lines []string) ([]model.DigestRow, error) {
	rows := make([]model.DigestRow, 0, len(lines))
	for _, line := range lines {
		if line == "" {
			continue
		}
		fields := strings.SplitN(line, "\t", 6)
		if len(fields) != 6 {
			return nil, fmt.Errorf("malformed DIGEST row %q", line)
		}
		values := make([]int, 4)
		names := []string{"count_star", "sum_time", "min_time", "max_time"}
		for index := range values {
			number, err := parseInt(fields[index+1], names[index])
			if err != nil {
				return nil, err
			}
			values[index] = number
		}
		rows = append(rows, model.DigestRow{
			Digest: fields[0], Count: values[0], SumTime: values[1],
			MinTime: values[2], MaxTime: values[3], Query: Sanitize(fields[5]),
		})
	}
	return rows, nil
}

func ParseBackend(lines []string) (model.BackendRows, error) {
	if len(lines) == 0 {
		return model.BackendRows{}, nil
	}
	poolIndex, pingIndex := -1, -1
	for index, line := range lines {
		switch line {
		case queries.BackendPoolMarker:
			poolIndex = index
		case queries.BackendPingMarker:
			pingIndex = index
		}
	}
	if poolIndex < 0 || pingIndex < 0 || poolIndex >= pingIndex {
		return model.BackendRows{}, errorsNew("BACKEND payload is missing or has invalid section markers")
	}
	var result model.BackendRows
	for _, line := range lines[poolIndex+1 : pingIndex] {
		if line == "" {
			continue
		}
		fields := strings.Split(line, "\t")
		if len(fields) != 7 {
			return model.BackendRows{}, fmt.Errorf("malformed BACKEND pool row %q", line)
		}
		values := make([]int, 4)
		for index, name := range []string{"ConnUsed", "ConnFree", "ConnOK", "ConnERR"} {
			number, err := parseInt(fields[index+3], name)
			if err != nil {
				return model.BackendRows{}, err
			}
			values[index] = number
		}
		result.Pool = append(result.Pool, model.PoolRow{
			Hostgroup: fields[0], ServerHost: fields[1], Status: fields[2],
			ConnUsed: values[0], ConnFree: values[1], ConnOK: values[2], ConnErr: values[3],
		})
	}
	for _, line := range lines[pingIndex+1:] {
		if line == "" {
			continue
		}
		fields := strings.SplitN(line, "\t", 4)
		if len(fields) != 4 {
			return model.BackendRows{}, fmt.Errorf("malformed BACKEND ping row %q", line)
		}
		var success *int
		if fields[2] != "" && !strings.EqualFold(fields[2], "NULL") {
			number, err := parseInt(fields[2], "ping_success_time_us")
			if err != nil {
				return model.BackendRows{}, err
			}
			success = &number
		}
		result.Ping = append(result.Ping, model.PingRow{
			Hostname: fields[0], LastPing: fields[1], SuccessUS: success,
			Error: Sanitize(fields[3]),
		})
	}
	return result, nil
}

func errorsNew(message string) error {
	return fmt.Errorf("%s", message)
}

func render(cleanLines, coloredLines []string, count int) model.RenderedView {
	return model.RenderedView{
		Clean: strings.Join(cleanLines, "\n"), Colored: strings.Join(coloredLines, "\n"),
		RowCount: count,
	}
}

func FormatConnections(
	rows, previous []model.ConnectionRow,
	threshold int,
	userFilter string,
) (model.RenderedView, error) {
	filter, err := CompileUserFilter(userFilter)
	if err != nil {
		return model.RenderedView{}, err
	}
	type key struct{ user, client, server, schema string }
	old := make(map[key]int, len(previous))
	for _, row := range previous {
		old[key{row.User, row.ClientHost, row.ServerHost, row.Schema}] = row.Connections
	}
	header := fmt.Sprintf("%-20s | %-15s | %-28s | %-20s | %-10s | %-10s",
		"USER", "SOURCE (Cli)", "BACKEND (Srv)", "SCHEMA", "CONN", "DELTA")
	clean, colored := []string{header}, []string{bold + header + reset}
	count := 0
	totalConnections, totalDelta := 0, 0
	for _, row := range rows {
		if !filter.MatchString(Sanitize(row.User)) {
			continue
		}
		rowKey := key{row.User, row.ClientHost, row.ServerHost, row.Schema}
		previousValue, exists := old[rowKey]
		delta := row.Connections - previousValue
		deltaText := "0"
		if !exists {
			delta = 0
		} else if delta != 0 {
			deltaText = fmt.Sprintf("%+d", delta)
		}
		totalConnections += row.Connections
		totalDelta += delta
		line := fmt.Sprintf("%-20s | %-15s | %-28s | %-20s | %-10d | %-10s",
			truncate(Sanitize(row.User), 20), truncate(Sanitize(row.ClientHost), 15),
			truncate(Sanitize(row.ServerHost), 28), truncate(Sanitize(row.Schema), 20),
			row.Connections, deltaText)
		color := green
		if threshold > 0 && row.Connections >= threshold {
			color = red
		}
		clean, colored = append(clean, line), append(colored, color+line+reset)
		count++
	}
	if totalConnections > 0 {
		deltaText := "0"
		if totalDelta != 0 {
			deltaText = fmt.Sprintf("%+d", totalDelta)
		}
		line := fmt.Sprintf("%-20s | %-15s | %-28s | %-20s | %-10d | %-10s",
			"GLOBAL TOTALS", "", "", "", totalConnections, deltaText)
		clean, colored = append(clean, line), append(colored, bold+cyan+line+reset)
	}
	return render(clean, colored, count), nil
}

func FormatQueries(rows []model.QueryRow, userFilter string, terminalWidth int) (model.RenderedView, error) {
	filter, err := CompileUserFilter(userFilter)
	if err != nil {
		return model.RenderedView{}, err
	}
	queryWidth := max(20, terminalWidth-97)
	header := fmt.Sprintf("%-8s | %-4s | %-15s | %-15s | %-28s | %-9s | %-*s",
		"PSID", "HG", "USER", "SOURCE", "BACKEND", "TIME", queryWidth, "ACTIVE QUERY")
	clean, colored := []string{header}, []string{bold + header + reset}
	count := 0
	for _, row := range rows {
		if !filter.MatchString(Sanitize(row.User)) {
			continue
		}
		line := fmt.Sprintf("%-8s | %-4s | %-15s | %-15s | %-28s | %-9s | %-*s",
			truncate(Sanitize(row.SessionID), 8), truncate(Sanitize(row.Hostgroup), 4),
			truncate(Sanitize(row.User), 15), truncate(Sanitize(row.ClientHost), 15),
			truncate(Sanitize(row.ServerHost), 28), fmt.Sprintf("%dms", row.TimeMS),
			queryWidth, truncate(Sanitize(row.Query), queryWidth))
		color := cyan
		if row.TimeMS > 1000 {
			color = red
		} else if row.TimeMS > 500 {
			color = yellow
		}
		clean, colored = append(clean, line), append(colored, color+line+reset)
		count++
	}
	return render(clean, colored, count), nil
}

func FormatDigests(rows []model.DigestRow, terminalWidth int) model.RenderedView {
	queryWidth := max(20, terminalWidth-73)
	header := fmt.Sprintf("%-18s | %-10s | %-10s | %-10s | %-10s | %-*s",
		"DIGEST", "COUNT", "SUM_TIME", "MIN_TIME", "MAX_TIME", queryWidth, "QUERY TEXT")
	clean, colored := []string{header}, []string{bold + header + reset}
	for _, row := range rows {
		line := fmt.Sprintf("%-18s | %-10d | %-10d | %-10d | %-10d | %-*s",
			truncate(Sanitize(row.Digest), 18), row.Count, row.SumTime, row.MinTime, row.MaxTime,
			queryWidth, truncate(Sanitize(row.Query), queryWidth))
		clean, colored = append(clean, line), append(colored, cyan+line+reset)
	}
	return render(clean, colored, len(rows))
}

func backendStatusColor(status string) string {
	switch strings.ToUpper(strings.TrimSpace(status)) {
	case "ONLINE":
		return green
	case "SHUNNED":
		return yellow
	case "OFFLINE_SOFT":
		return orange
	case "OFFLINE_HARD":
		return red
	default:
		return red
	}
}

func FormatBackend(rows model.BackendRows) model.RenderedView {
	header := fmt.Sprintf("%-10s | %-35s | %-15s | %-11s | %-11s | %-11s | %-11s",
		"HOSTGROUP", "BACKEND HOST", "STATUS", "CONN USED", "CONN FREE", "CONN OK", "CONN ERR")
	clean, colored := []string{header}, []string{bold + header + reset}
	for _, row := range rows.Pool {
		line := fmt.Sprintf("%-10s | %-35s | %-15s | %-11d | %-11d | %-11d | %-11d",
			truncate(Sanitize(row.Hostgroup), 10), truncate(Sanitize(row.ServerHost), 35),
			truncate(Sanitize(row.Status), 15),
			row.ConnUsed, row.ConnFree, row.ConnOK, row.ConnErr)
		color := backendStatusColor(row.Status)
		clean, colored = append(clean, line), append(colored, color+line+reset)
	}
	pingHeader := fmt.Sprintf("%-35s | %-25s | %-15s | %s",
		"HOSTNAME", "LAST PING DATETIME", "SUCCESS (us)", "PING ERROR")
	clean, colored = append(clean, pingHeader), append(colored, bold+pingHeader+reset)
	for _, row := range rows.Ping {
		success := ""
		if row.SuccessUS != nil {
			success = strconv.Itoa(*row.SuccessUS)
		}
		line := fmt.Sprintf("%-35s | %-25s | %-15s | %s",
			truncate(Sanitize(row.Hostname), 35), truncate(Sanitize(row.LastPing), 25),
			success, Sanitize(row.Error))
		normalizedError := strings.ToUpper(strings.TrimSpace(row.Error))
		color := green
		if normalizedError != "" && normalizedError != "NULL" {
			color = red
		}
		clean, colored = append(clean, line), append(colored, color+line+reset)
	}
	return render(clean, colored, len(rows.Pool)+len(rows.Ping))
}
